src/status_im2/contexts/add_new_contact/events.cljs

```diff
@@ -6,114 +6,133 @@
+  (let [regex         #"^https?://join.status.im/u/(.+)"
+        trimmed-input (utils/safe-trim input)]
+    (->> {:id (if (empty? trimmed-input)
+                nil
+                (if-some [[_ id] (re-matches regex trimmed-input)]
+                  id
+                  trimmed-input))}
+         (merge contact))))
+
+(defn ->type
+  [{:keys [id] :as contact}]
+  (->> (cond
+         (empty? id)                  {:type :empty}
+         (v/valid-public-key? id)     {:type       :public-key
+                                       :public-key id}
+         (v/valid-compressed-key? id) {:type :compressed-key}
+         :else                        {:type :ens
+                                       :ens  (stateofus/ens-name-parse id)})
+       (merge contact)))
+
+(defn ->state
+  [{:keys [id type public-key user-public-key] :as contact}]
+  (->> (cond
+         (empty? id)                            {:state :empty}
+         (= type :public-key)                   {:state :invalid
+                                                 :msg   :t/not-a-chatkey}
+         (= public-key user-public-key)         {:state :invalid
+                                                 :msg   :t/can-not-add-yourself}
+         (and (= type :compressed-key)
+              (empty? public-key))              {:state :decompress-key}
+         (and (= type :ens)
+              (empty? public-key))              {:state :resolve-ens}
+         (and (or (= type :compressed-key) (= type :ens))
+              (v/valid-public-key? public-key)) {:state :valid})
+       (merge contact)))
+
+(def validate-contact (comp ->state ->type ->id))
+
+(defn cb [event contact] (fn [arg] (rf/dispatch [event contact arg])))
+
+(rf/defn set-new-identity
+  {:events [:contacts/set-new-identity]}
+  [{:keys [db]} input]
+  (let [upk             (get-in db [:multiaccount :public-key])
+        init            (init-contact {:user-public-key upk
+                                       :input           input})
+        {:keys [id ens state]
+         :as   contact} (validate-contact init)]
+    (case state
+      :empty            {:db (dissoc db :contacts/new-identity)}
+      (:valid :invalid) {:db (assoc db :contacts/new-identity contact)}
+      :decompress-key   {:db (assoc db :contacts/new-identity contact)
+                         :contacts/decompress-public-key
+                         {:compressed-key id
+                          :on-success
+                          (cb :contacts/set-new-identity-success contact)
+                          :on-error
+                          (cb :contacts/set-new-identity-error contact)}}
+      :resolve-ens      {:db (assoc db :contacts/new-identity contact)
+                         :contacts/resolve-public-key-from-ens
+                         {:chain-id (ethereum/chain-id db)
+                          :ens ens
+                          :on-success
+                          (cb :contacts/set-new-identity-success contact)
```

Comment:
`#(rf/dispatch [:contacts/set-new-identity-success contact %])` simpler and more readable

Reply:
Thanks for your review @flexsurfer!

During these PR's, I always wonder whether I should implement requested changes first, and then push that along with comments at the same time (because sometimes my opinions/decisions change when I'm implementing them and if I were to comment first, I'd have to update the comments). This time I'll just do a first-pass on comments to explain reasoning and ideas, before actually implementation.

Your comment got me really excited because I realized I might not actually need that `cb` function after all - I can probably just `with-redef` the `rf/dispatch` directly. That part (being able to replace dispatch with an identity-kind of function) is important for testing the handlers and is the part I like the most (I've been thinking about how to test event handlers and their callbacks, along with others on the @ilmotta @yqrashawn), so I'll try what you're suggesting, and if it works, I think it will be way better. Thanks!

Reply:
@flexsurfer I ended up having to keep it (renamed to `dispatcher`, following @ilmotta's suggestion). When called by the test, the mocked `dispatcher` function `(fn [& args] args)` merges arg text with the returned map, which can be asserted in the test, whereas the production-code dispatcher function returns a function that dispatches the args.